lib/logdna/client.rb

```diff
@@ -1,164 +1,137 @@
+      @flush_interval = opts[:flush_interval] || Resources::FLUSH_INTERVAL
+      @flush_scheduled = false
+      @exception_flag = false
+
+      @request = request
+      @retry_timeout = opts[:retry_timeout] || Resources::RETRY_TIMEOUT
     end
 
-    def message_hash(msg, opts={})
-      obj = {
+    def process_message(msg, opts = {})
+      processed_message = {
         line: msg,
         app: opts[:app],
         level: opts[:level],
         env: opts[:env],
         meta: opts[:meta],
         timestamp: Time.now.to_i,
       }
-      obj.delete(:meta) if obj[:meta].nil?
-      obj
+      processed_message.delete(:meta) if processed_message[:meta].nil?
+      processed_message
     end
 
-    def create_flush_task
-      return @task unless @task.nil? or !@task.running?
-
-      t = Concurrent::TimerTask.new(execution_interval: @actual_flush_interval, timeout_interval: Resources::TIMER_OUT) do |task|
-        if @messages.any?
-          # keep running if there are queued messages, but don't flush
-          # because the buffer is being flushed due to being over the limit
-          unless @buffer_over_limit
-            flush()
-          end
-        else
-          # no messages means we can kill the task
-          task.kill
-        end
-      end
-      t.execute
-    end
-
-    def check_side_buffer
-      return if @side_buffer.size == 0
-
-      @buffer.write(@side_buffer.string)
-      @side_buffer.truncate(0)
-      queued_side_messages = @side_messages
-      @side_messages = []
-      queued_side_messages.each { |message_hash_obj| @messages.push(message_hash_obj) }
+    def schedule_flush
+      start_timer = lambda {
+        sleep(@exception_flag ? @retry_timeout : @flush_interval)
+        flush if @flush_scheduled
+      }
+      thread = Thread.new { start_timer }
+      thread.join
     end
 
-
-    # this should always be running synchronously within this thread
-    def buffer(msg, opts)
-      buffer_size = write_to_buffer(msg, opts)
-      unless buffer_size.nil?
-        process_buffer(buffer_size)
+    def write_to_buffer(msg, opts)
+      if @lock.try_lock
+        processed_message = process_message(msg, opts)
+        new_message_size = processed_message.to_s.bytesize
+        @buffer.push(processed_message)
+        @buffer_byte_size += new_message_size
+        @flush_scheduled = true
+        @lock.unlock
+
+        flush if @flush_limit <= @buffer_byte_size
+        schedule_flush unless @flush_scheduled
+      else
+        @side_message_lock.synchronize do
+          @side_messages.push(process_message(msg, opts))
+        end
       end
     end
 
-    def write_to_buffer(msg, opts)
-      return if msg.nil?
-      msg = encode_message(msg)
-
-      if @lock.locked?
-        @side_buffer.write(msg)
-        @side_messages.push(message_hash(msg, opts))
-        return
+    def send_request
+      @side_message_lock.synchronize do
+        @buffer.concat(@side_messages)
```

Comment:
I think we should leave a comment here that this function must be called with a claimed `@lock` (and/or even explicitly check for that in the code). Otherwise when I read this I'm confused why we're modifying `@buffer` without looking at `@lock`.